File: olly-nodeJs/src/colour.js
import {tuple} from './tuple'
import d       from './utils/decimal'

const colour = (...args) => ([...args])

const red = t => t[0]
const green = t => t[1]
const blue = t => t[2]

const multiply = (t1, t2) => {
  return tuple(
    d(red(t1)).times(d(red(t2))).toNumber(),
    d(green(t1)).times(d(green(t2))).toNumber(),
    d(blue(t1)).times(d(blue(t2))).toNumber(),
  )
}

const toBytes = c => (c.map(num => (255*num)))

const fromBytes = c => (c.map(num => (num/255)))

export {colour, red, green, blue, multiply, toBytes, fromBytes}

File: tim-rails/app/models/colour.rb
# class Colour < Tuple
#   attr_accessor :x,
#                 :y,
#                 :z
#
#   def red
#     x
#   end
#
#   def green
#     y
#   end
#
#   def blue
#     z
#   end
#
#   def add(colour)
#     result = Colour.new
#     result.x = x + colour.x
#     result.y = y + colour.y
#     result.z = z + colour.z
#     result
#   end
#
#   def subtract(colour)
#     result = Colour.new
#     result.x = x - colour.x
#     result.y = y - colour.y
#     result.z = z - colour.z
#     result
#   end
#
#   def multiply(scalar)
#     result = Colour.new
#     result.x = x.to_f * scalar
#     result.y = y.to_f * scalar
#     result.z = z.to_f * scalar
#     result
#   end
#
#   def multiply_by_colour(colour)
#     result = Colour.new
#     result.x = x * colour.x
#     result.y = y * colour.y
#     result.z = z * colour.z
#     result
#   end
# end
